src/libcompiler_builtins/probestack.rs

```diff
@@ -0,0 +1,153 @@
+//! don't have to maintain assembly for MSVC and MinGW, we can just have one
+//! blob of assembly.
+//!
+//! Finally it's worth noting that at the time of this writing LLVM only has
+//! support for stack probes on x86 and x86_64. There's no support for stack
+//! probes on any other architecture like ARM or PowerPC64. LLVM I'm sure would
+//! be more than welcome to accept such a change!
+
+use core::intrinsics;
+
+#[naked]
+#[no_mangle]
+#[cfg(target_arch = "x86_64")]
+pub unsafe extern fn __rust_probestack() {
+    // our goal here is to take %eax and add it to %rsp, but we're also going to
+    // touch each page between %rsp+8 and %rsp+8-%rax
+    //
+    // The ABI here is that the stack frame size is located in `%eax`. Upon
+    // return we're not supposed to modify `%esp` or `%eax`. It's unclear
+    // whether we can modify caller-saved registers, but to be on the safe side
+    // our one scratch register here, `%rcx`, is preserved in this function.
+    asm!("
+        push   %rcx
+        lea    16(%rsp),%rcx     // rsp before calling this routine -> rcx
+
+        // Main loop, taken in one page increments. We're decrementing rcx by
+        // a page each time until there's less than a page remaining. We're
+        // guaranteed that this function isn't called unless there's more than a
+        // page needed
+    2:
+        sub    $$0x1000,%rcx
+        test   %rcx,(%rcx)
+        sub    $$0x1000,%rax
+        cmp    $$0x1000,%rax
+        ja     2b
+
+        // Finish up the last remainings stack space requested, getting the last
+        // bits out of rax
+        sub    %rax,%rcx
+        test   %rcx,(%rcx)
+
+        // We now know that %rcx is (%rsp + 16 - %rax) so to recover rax
+        // we calculate (%rsp + 16) - %rcx which will give us %rax
+        lea    16(%rsp),%rax
+        sub    %rcx,%rax
+
+        // Restore %rcx and leave the function
+        pop    %rcx
+        ret
+    ");
+    intrinsics::unreachable();
+}
+
+#[naked]
+#[no_mangle]
+#[cfg(all(target_arch = "x86", unix))]
+pub unsafe extern fn __rust_probestack() {
+    // This is the same as x86_64 above, only translated for 32-bit sizes. Note
+    // that on Unix we're expected to restore everything as it was, this
+    // function basically can't tamper with anything.
+    //
+    // The ABI here is the same as x86_64, except everything is 32-bits large.
+    asm!("
+        push   %ecx
+        lea    8(%esp),%ecx
+    2:
+        sub    $$0x1000,%ecx
+        test   %ecx,(%ecx)
+        sub    $$0x1000,%eax
+        cmp    $$0x1000,%eax
+        ja     2b
+
+        sub    %eax,%ecx
+        test   %ecx,(%ecx)
+
+        lea    8(%esp),%eax
+        sub    %ecx,%eax
+        pop    %ecx
+        ret
+    ");
+    intrinsics::unreachable();
+}
+
+#[naked]
+#[no_mangle]
+#[cfg(all(target_arch = "x86", windows))]
+pub unsafe extern fn __rust_probestack() {
+    // This is similar to 32-bit unix but we're going to actually perform the
```

Comment:
I'm pretty sure you don't need this function: just omit the `probe-stack` attribute on Windows and LLVM will automatically add a call to Microsoft's implementation. It's mandated by the Windows ABI, so LLVM has to do that.